File: app/src/main/java/es/etologic/mahjongscoring2/app/combinations/CombinationsViewModel.java
package es.etologic.mahjongscoring2.app.combinations;

import android.arch.lifecycle.LiveData;
import android.arch.lifecycle.MutableLiveData;

import java.util.ArrayList;
import java.util.List;

import es.etologic.mahjongscoring2.app.base.BaseViewModel;
import es.etologic.mahjongscoring2.domain.entities.Combination;
import es.etologic.mahjongscoring2.domain.threading.UseCase;
import es.etologic.mahjongscoring2.domain.threading.UseCaseHandler;
import es.etologic.mahjongscoring2.domain.use_cases.GetCombinationsUseCase;
import es.etologic.mahjongscoring2.domain.use_cases.GetFilteredCombinationsUseCase;

import static es.etologic.mahjongscoring2.app.model.ShowState.HIDE;
import static es.etologic.mahjongscoring2.app.model.ShowState.SHOW;

class CombinationsViewModel extends BaseViewModel {

    private final GetCombinationsUseCase getCombinationsUseCase;
    private GetFilteredCombinationsUseCase getFilteredCombinationsUseCase;
    private MutableLiveData<List<Combination>> combinations = new MutableLiveData<List<Combination>>() {};

    CombinationsViewModel(UseCaseHandler useCaseHandler,
                          GetCombinationsUseCase getCombinationsUseCase,
                          GetFilteredCombinationsUseCase getFilteredCombinationsUseCase) {
        super(useCaseHandler);
        this.getCombinationsUseCase = getCombinationsUseCase;
        this.getFilteredCombinationsUseCase = getFilteredCombinationsUseCase;
    }

    LiveData<List<Combination>> getCombinations() {
        return combinations;
    }

    void loadCombinations() {
        progressState.setValue(SHOW);
        useCaseHandler.execute(getCombinationsUseCase, null,
                new UseCase.UseCaseCallback<GetCombinationsUseCase.ResponseValue>() {
                    @Override
                    public void onSuccess(GetCombinationsUseCase.ResponseValue response) {
                        combinations.setValue(response.getCombinations());
                        progressState.setValue(HIDE);
                    }

                    @Override
                    public void onError(String ignored) {
                        combinations.setValue(new ArrayList<>());
                        progressState.setValue(HIDE);
                    }
                });
    }

    void searchCombination(String text) {
        useCaseHandler.execute(getFilteredCombinationsUseCase,
                new GetFilteredCombinationsUseCase.RequestValues(text),
                new UseCase.UseCaseCallback<GetFilteredCombinationsUseCase.ResponseValue>() {
                    @Override
                    public void onSuccess(GetFilteredCombinationsUseCase.ResponseValue response) {
                        combinations.setValue(response.getCombinations());
                        progressState.setValue(HIDE);
                    }

                    @Override
                    public void onError(String ignored) {
                        combinations.setValue(new ArrayList<>());
                        progressState.setValue(HIDE);
                    }
                });
    }
}

File: app/src/main/java/es/etologic/mahjongscoring2/app/main/combinations/CombinationsViewModel.kt
package es.etologic.mahjongscoring2.app.main.combinations

import androidx.lifecycle.LiveData
import androidx.lifecycle.MutableLiveData
import es.etologic.mahjongscoring2.app.base.BaseViewModel
import es.etologic.mahjongscoring2.app.model.ShowState.HIDE
import es.etologic.mahjongscoring2.app.model.ShowState.SHOW
import es.etologic.mahjongscoring2.domain.model.Combination
import es.etologic.mahjongscoring2.domain.use_cases.GetCombinationsUseCase
import io.reactivex.schedulers.Schedulers

internal class CombinationsViewModel(
    private val getCombinationsUseCase: GetCombinationsUseCase
) : BaseViewModel() {
    
    private val filteredCombinations = MutableLiveData<List<Combination>>()
    
    fun getFilteredCombinations(): LiveData<List<Combination>> {
        return filteredCombinations
    }
    
    fun getAll() {
        disposables.add(
            getCombinationsUseCase.getAll()
                .subscribeOn(Schedulers.io())
                .doOnSubscribe { progressState.postValue(SHOW) }
                .doOnEvent { _, _ -> progressState.postValue(HIDE) }
                .subscribe(filteredCombinations::postValue, error::postValue)
        )
    }
    
    fun searchCombination(filter: String) {
        disposables.add(
            getCombinationsUseCase.getSome(filter)
                .subscribeOn(Schedulers.io())
                .doOnSubscribe { progressState.postValue(SHOW) }
                .doOnEvent { _, _ -> progressState.postValue(HIDE) }
                .subscribe(filteredCombinations::postValue, error::postValue)
        )
    }
}
